napari/layers/labels/labels.py

```diff
@@ -555,26 +552,45 @@ def _set_colormap(self, colormap):
             else:
                 color_mode = LabelColorMode.DIRECT
                 self._colormap = self._direct_colormap
+        self._cached_labels = None  # invalidate the cached color mapping
         self._selected_color = self.get_color(self.selected_label)
+        self._color_mode = color_mode
         self.events.colormap()  # Will update the LabelVispyColormap shader
-        self.color_mode = color_mode
+        self.events.selected_label()
+        self.refresh()
 
     @property
     def num_colors(self):
         """int: Number of unique colors to use in colormap."""
-        return self._num_colors
+        warnings.warn(
+            trans._(
+                'Labels.num_colors is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0, please use len(Labels.colormap) '
```

Comment:
@psobolewskiPhD maybe we can discuss this in a follow-up issue or PR? I am of the opposite opinion, because we have way too many attributes in our objects — frequently when using napari I do `obj.<TAB>` and then I get an enormous list in which it is really hard to find what I'm looking for. I'm not sure whether colormaps fall in this category, but the point is that *eventually* all objects land in this category if you keep expanding the API for every thing that you want to know about. And, it's also real easy to keep adding things, but hard to remove them, so, especially while we are working to simplify our APIs, for every attribute we add to a class, I want to have a nice, long discussion about it. 😅 For standard Python things like "how many things are in this collection", I prefer to use `len()`. I realise that not everyone would think to do this, which is why I'm happy to contemplate in a follow-up issue, but since adding to the API is easy and not time-sensitive, I suggest we leave the discussion until post-0.5.

Reply:
fair enough